### PR TITLE
[Boundary Condition] Cohesion BC

### PR DESCRIPTION
**Pull request by LDTalbot**
**Describe the PR**
This PR introduces a cohesion boundary condition.

Progress Checklist:
- [x] Implement basic functionality
- [x] Write tests
- [x] Test with several variations of sliding block benchmark

Expanded Functionality Checklist:
- [x] Cartesian and rotated
- [x] 2D and 3D
- [x] Allow input via entity sets or file

**Related Issues/PRs**
Suggested for related future PR (can also apply to other BCs):
- Expand available element types
    - Write function to distinguish element types
    - Area calculations for non-square-faced elements
- Auto-calculation of node position

**Additional context**

### CODE REVIEW

**Comment by bodhinandach** [comment]
Thanks @LDTalbot! I will have a look at the PR this weekend.

File: include/io/io_mesh_ascii.tcc [hunk 1]
```diff
@@ -691,6 +691,62 @@ std::vector<std::tuple<mpm::Index, unsigned, int, double>>
...
+          // ID
+          mpm::Index id;
+          // Direction
+          unsigned dir;
+          // Sign
+          int sign;
+          // Cohesion
+          double cohesion;
+          // Cell height
+          double h_min;
```
Comment (bodhinandach) [inline include/io/io_mesh_ascii.tcc:728]: I am not sure I understand why we need to define `h_min` in the input for every node. In the current stage, we only use a constant h_min for all elements, is that possible to use the function `compute_average_cell_size()`?

File: include/io/io_mesh_ascii.tcc [hunk 2]
```diff
@@ -691,6 +691,62 @@ std::vector<std::tuple<mpm::Index, unsigned, int, double>>
...
+          // Node nposition
+          int nposition;  // LEDT TODO not reading? use #s in input file?
```
Comment (bodhinandach) [inline include/io/io_mesh_ascii.tcc:730]: What is this used for? Is that to indicate the ID for the nodal position? Also, remove LEDT once you are done with debugging.

File: include/nodes/node.tcc [hunk 3]
```diff
@@ -654,6 +656,233 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
+
+    // Set cohesion constraint
+    // Direction value in the constraint (0, Dim)
+    const unsigned dir_n = std::get<0>(this->cohesion_constraint_);
+
+    // Normal direction of cohesion
+    const double sign_dir_n = sign(std::get<1>(this->cohesion_constraint_));
+
+    // Cohesion or undrained shear strength
+    const double su = std::get<2>(this->cohesion_constraint_);
```
Comment (bodhinandach) [inline include/nodes/node.tcc:697]: Maybe it is good to make the variable naming constant.

File: include/nodes/node.tcc [hunk 4]
```diff
@@ -654,6 +656,233 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
+    const unsigned phase = 0;
+
+    // Acceleration and velocity
+    double acc_n, acc_t, vel_n, vel_t;
+
+    // Nodal area
+    double nodal_area_;
+
+    // Calculate area associated with node
+    try {
```
Comment (bodhinandach) [inline include/nodes/node.tcc:714]: I think you can remove the try-catch block here.

File: include/nodes/node.tcc [hunk 5]
```diff
@@ -654,6 +656,233 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
+
+      if (acc_n * sign_dir_n > 0.0) {
+        // kinetic
+        if (vel_t != 0.0) {
+          Eigen::Matrix<double, 2, 1> vel_net;
+          // cohesion is applied opposite to the vel_net
+          vel_net(0) = vel(dir_t0) + acc(dir_t0) * dt;
+          vel_net(1) = vel(dir_t1) + acc(dir_t1) * dt;
+          const double vel_net_t =
+              sqrt(vel_net(0) * vel_net(0) + vel_net(1) * vel_net(1));
```
Comment (bodhinandach) [inline include/nodes/node.tcc:847]: Can we use `.norm()` here?

File: include/solvers/mpm_base.tcc [hunk 6]
```diff
@@ -1139,6 +1142,64 @@ void mpm::MPMBase<Tdim>::nodal_frictional_constraints(
...
+          // Set id
+          int nset_id = constraints.at("nset_id").template get<int>();
+          // Direction
+          unsigned dir = constraints.at("dir").template get<unsigned>();
+          // Sign n
+          int sign_n = constraints.at("sign_n").template get<int>();
+          // Cohesion
+          double cohesion = constraints.at("cohesion").template get<double>();
+          // h_min
+          double h_min = constraints.at("h_min").template get<double>();
```
Comment (bodhinandach) [inline include/solvers/mpm_base.tcc:1180]: I would use compute_average_cell_size() if it is not defined.

**Review by bodhinandach — COMMENTED** [review]
@LDTalbot Thanks for the PR. I have looked into them and they are implemented nicely. I have some suggestions which hopefully can be considered.

File: include/io/io_mesh_ascii.tcc (hunk 2 above)
Comment (LDTalbot) [reply include/io/io_mesh_ascii.tcc]: Yes, temporary input variable for whether the node is corner, edge, face w.r.t. mesh bounding box for area calculations (will be removed once area calculations are implemented).

I have removed the extra debugging comments also.

**Comment by bodhinandach** [comment]
Once everything is fixed, can you pass me one of your simulation files? I want to test if the parallel processes work with the code or not.

File: include/io/io_mesh_ascii.tcc (hunk 1 above)
Comment (LDTalbot) [reply include/io/io_mesh_ascii.tcc:728]: Not implemented for now because cohesion is initialized before mesh/cells.

File: include/nodes/node.tcc (hunk 3 above)
Comment (LDTalbot) [reply include/nodes/node.tcc:697]: Have cleaned up variable naming (did some cleanup in friction constraint too while I was at it). The only places where I switch a couple variable names now is for convenience within the apply_cohesion_constraints() implementation (node.tcc).

File: include/nodes/node.tcc (hunk 4 above)
Comment (LDTalbot) [reply include/nodes/node.tcc:714]: Moved/combined with try-catch block in assign_cohesion_constraints(), and updated tests to match where this function will have status = false.

File: include/nodes/node.tcc (hunk 5 above)
Comment (LDTalbot) [reply include/nodes/node.tcc:847]: Done

File: include/solvers/mpm_base.tcc (hunk 6 above)
Comment (LDTalbot) [reply include/solvers/mpm_base.tcc:1180]: See other comment

**Comment by jgiven100** [comment]
@LDTalbot I took a look at the private variable naming - everything looks good to me (but I'm not necessarily the most qualified person to check based on how many slipped through the absorbing boundary constraint 😄 )!

**Review by bodhinandach — CHANGES_REQUESTED** [review]
@LDTalbot I checked that this branch works ok with MPI. I have some minor requests, but otherwise, this PR should be working.

File: include/nodes/node.tcc [hunk 7]
```diff
@@ -654,6 +656,228 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
+            throw std::runtime_error("Invalid cohesion boundary nposition");
+            break;
+        }
+      }
+
+      // Assign to tuple
+      this->cohesion_constraint_ = std::make_tuple(
+          static_cast<unsigned>(dir), static_cast<int>(sign_n),
+          static_cast<double>(cohesion), static_cast<double>(nodal_area),
+          static_cast<int>(nposition));
```
[the remark this answers is not in the document]
Comment (LDTalbot) [reply include/nodes/node.tcc]: Good catch, thank you. Have addressed this in latest commit.

File: include/nodes/node.tcc [hunk 8]
```diff
@@ -654,6 +656,228 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
+    const double sign_n = sign(std::get<1>(this->cohesion_constraint_));
+
+    // Cohesion or undrained shear strength
+    const double su = std::get<2>(this->cohesion_constraint_);
+
+    // Nodal area
+    const double nodal_area = std::get<3>(this->cohesion_constraint_);
+
+    // Location of node for area computation
+    const int nposition = std::get<4>(this->cohesion_constraint_);
```
[the remark this answers is not in the document]
Comment (LDTalbot) [reply include/nodes/node.tcc]: Addressed in latest commit 🙂

**Review by bodhinandach — APPROVED** [review]
Let's merge this after the tests are done! Thank you!

**Comment by bodhinandach** [comment]
@jgiven100 @yliang-sn Please have a quick look at this PR before the meeting today if you can.

**Comment by jgiven100** [comment]
@bodhinandach Yes - will do!

File: include/loads_bcs/cohesion_constraint.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  // Direction (normal)
+  unsigned dir_;
+  // Sign of normal direction
+  int sign_n_;
+  // Cohesion
+  double cohesion_;
+  // Cell height
+  double h_min_;
+  // Node nposition
+  int nposition_;
```
Comment (jgiven100) [inline include/loads_bcs/cohesion_constraint.h:63]: Above mentions that `nposition_` default is `0`. It is not initialized as zero here. I don't see its assigned anywhere else.

It seems if I set `nposotion_ = 5`, the switch statement below throws an error. So there is no need for a no-position specific value? Am I following this correctly?

**Review by jgiven100 — APPROVED** [review]
I am ok with this. 1 question regarding `nposition_` default.

File: include/loads_bcs/cohesion_constraint.h (hunk 9 above)
Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: I think the error will be thrown. We probably can remove the comment "_0: None (no nposition is specified)_" to make it clearer.

Comment (jgiven100) [reply include/loads_bcs/cohesion_constraint.h:63]: I think the only scenario where anything would come up is if somebody is reading constraints from a file and forgets to include the last column of positional `int`s. 

```
istream >> id >> dir >> sign_n >> cohesion >> h_min >> nposition;
```

I think the istream extractor will just not re-assign nposition which is initialized without a value above. So maybe there is chance that the nposition value is accidentally 1, 2, or 3.

File: include/io/io_mesh_ascii.tcc [hunk 10]
```diff
@@ -691,6 +692,62 @@ std::vector<std::tuple<mpm::Index, unsigned, int, double>>
...
+          int sign_n;
+          // Cohesion
+          double cohesion;
+          // Cell height
+          double h_min;
+          // Node nposition
+          int nposition;
+          while (istream.good()) {
+            // Read stream
+            istream >> id >> dir >> sign_n >> cohesion >> h_min >> nposition;
```
Comment (jgiven100) [inline include/io/io_mesh_ascii.tcc:734]: @bodhinandach here

File: include/loads_bcs/cohesion_constraint.h (hunk 9 above)
Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: I am not sure if someone forget to add the last column. I guess three things might happen: (i) `std::istream` will throw, (ii) wont read anything, (iii) might read the next line first column (unlikely). If (i) it is good, since the user is warned due to incompatible input. If (ii) it is ok since if `nposition` is not 1, 2, or 3, the error message at `node.tcc` will be thrown. If (iii), then this might be an issue, since the next line first element is `id`, which is also an unsigned/int. Is this answering your concern, @jgiven100? @LDTalbot can you give it a try?

Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: Oh, I am not sure the default value will be 1,2,or 3. It will be most likely a random "garbage" value.

Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: https://www.learncpp.com/cpp-tutorial/uninitialized-variables-and-undefined-behavior/

Comment (LDTalbot) [reply include/loads_bcs/cohesion_constraint.h:63]: Testing this now (cohesion input file with no nposition column), looks like so far it runs so making it output nposition to see what is happening. 

Also note that the default value in cohesion_constraint.h is 0 but Joel doesn't think that matters since is already passed.

Comment (LDTalbot) [reply include/loads_bcs/cohesion_constraint.h:63]: When using file input and neglecting a column, it's spitting out nposition = 0 and throwing "Invalid cohesion boundary nposition"

Comment (jgiven100) [reply include/loads_bcs/cohesion_constraint.h:63]: @bodhinandach Yes. As indicated in [this comment](https://github.com/geomechanics/mpm/pull/65#discussion_r966373301), it seems `istream >>` does what you call option (ii) when the last column is missing. The "garbage" initialization value is what is passed along in the code. I agree it seems very unlikely the "garbage" value will just so happen to be 1, 2, or 3,

@LDTalbot will push 
```
int nposition = 0;
```
to resolve.

Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: So it seems that it should be OK...

Comment (LDTalbot) [reply include/loads_bcs/cohesion_constraint.h:63]: Joel thinks this may be luck/'istream' behaviour, pushing a commit with L731 of io_mesh_ascii.tcc as `int nposition = 0;` to make sure neglecting nposition column in file input throws error.

Comment (LDTalbot) [reply include/loads_bcs/cohesion_constraint.h:63]: note: if missing in entity sets always throws error with "... key 'nposition' not found"

Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: I am not sure we need to check it all the time we read. That's going to cost some time. Other read functions are not having this too. The one implemented in node.tcc should serve as a check already.

Comment (jgiven100) [reply include/loads_bcs/cohesion_constraint.h:63]: Theres no additional checks. Just guaranteed to have `nposition=0` for the case that I mess up input files.

Comment (bodhinandach) [reply include/loads_bcs/cohesion_constraint.h:63]: Sounds good to me. Let's put:
`int nposition = 0;`

**Review by bodhinandach — APPROVED** [review]

**Review by jgiven100 — APPROVED** [review]

**Review by yliang-sn — APPROVED** [review]
I am OK with the current structure.

**Review by bodhinandach — APPROVED** [review]
